### PR TITLE
applications: nrf5340_audio: Implement RGB Led Music Sync

### PR DESCRIPTION
**Pull request by Khuongh**
Implementing music sync for RGB Led

Signed-off-by: Khuong Huynh <khuong.huynh@nordicsemi.no>

### CODE REVIEW

**Comment by Khuongh** [comment]
CPU usage of Music Color Sync Thread

Music_color_sync_cpu_stack

File: applications/nrf5340_audio/src/main.c [hunk 1]
```diff
@@ -233,6 +234,13 @@ void main(void)
 	ret = streamctrl_start();
 	ERR_CHK(ret);
 
+#if (CONFIG_AUDIO_DEV == HEADSET && CONFIG_MUSIC_COLOR_SYNC) 
+	ret = music_color_sync_init();
+	if (ret) {
+		LOG_ERR("Failed to init music color sync - (err: %d)", ret);
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/main.c:240]:
```suggestion
		LOG_ERR("Failed to init music color sync, ret: %d", ret);
```

File: applications/nrf5340_audio/src/main.c [hunk 2]
```diff
@@ -233,6 +234,13 @@ void main(void)
...
+	}
+#endif
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/main.c:242]: Add endif comment

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 3]
```diff
@@ -49,6 +49,53 @@ config I2S_CH_NUM
...
 endmenu # I2S
 
+#----------------------------------------------------------------------------#
+menu "Music Color Synchronization"
+
+config MUSIC_COLOR_SYNC
+	bool "Enabling RGB synchronization with music"
+	default n
+	help
+		"NRFX_PWM0" and "CMSIS_DSP_TRANSFORM" has to be enabled
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/Kconfig:59]: Probably better to add "depends on" here

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 4]
```diff
@@ -49,6 +49,53 @@ config I2S_CH_NUM
...
+		for this functionality to work.
+
+menu "FFT"
+visible if MUSIC_COLOR_SYNC
+
+config FFT_SAMPLE_SIZE
+	int "Sample size for FFT"
+	default 256
+	help
+		 Supported FFT Lengths are 32, 64, 128, 256, 512,
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/Kconfig:69]:
```suggestion
		 Supported FFT sample sizes are 32, 64, 128, 256, 512,
```

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 5]
```diff
@@ -49,6 +49,53 @@ config I2S_CH_NUM
...
+		0 for FFT and 1 for IFFT
+
+config FFT_FLAG_BIT_REVERSAL
+	int "Flag for bit reversal of output"
+	range 0 1
+	default 1
+	help
+		0 to disable and 1 to enable bitreversal
+
+config RGB_RED_FREQ_BIN_RANGE
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/Kconfig:86]: Why are these needed?

Also, there is no config for blue?

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 6]
```diff
@@ -49,6 +49,53 @@ config I2S_CH_NUM
...
+	int "Freq range for RED RGB LED"
+	range 0 256
+	default 22 
+
+config RGB_GREEN_FREQ_BIN_RANGE
+	int "Freq range for GREEN RGB LED"
+	range 0 256
+	default 100
+
+endmenu # Music sync
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/Kconfig:96]: Hmm these endmenu comments does not match?

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:1]: Must have license text here

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+#ifndef _MUSIC_COLOR_SYNC_H_
+#define _MUSIC_COLOR_SYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**@brief Updates decoded pcm data used for FFT
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:7]: I think it is best to have @brief on the next line

Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:7]:
```suggestion
/**@brief Update PCM data used for FFT 
```

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ */
+int music_color_sync_data_update(char *pcm_data_mono, uint16_t len);
+
+/**@brief Initialize the music color synchronization
+ * 
+ * @return 0 if success, error otherwise
+ */
+int music_color_sync_init();
+
+#endif /* _MUSIC_COLOR_SYNC_H */
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:22]: Must have a newline at the end of every file

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+
+#ifndef _MUSIC_COLOR_SYNC_H_
+#define _MUSIC_COLOR_SYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**@brief Updates decoded pcm data used for FFT 
+ * 
+ * @param[in] pcm_data_mono pointer to decoded pcm data (mono)
+ * @param[in] len Size of decoded data
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:10]: Dont need [in] here since it is not a pointer

Also, it is strange that the variable name is len, but explanation says size. Should say lenght

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 11]
```diff
@@ -0,0 +1,22 @@
+
+#ifndef _MUSIC_COLOR_SYNC_H_
+#define _MUSIC_COLOR_SYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**@brief Updates decoded pcm data used for FFT 
+ * 
+ * @param[in] pcm_data_mono pointer to decoded pcm data (mono)
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:9]:
```suggestion
 * @param[in] pcm_data_mono Pointer to PCM data (mono)
```

File: applications/nrf5340_audio/src/modules/music_color_sync.h [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ * @param[in] len Size of decoded data
+ *
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/modules/music_color_sync.h:11]: Should have an explanation here of how the data is used inside the module

**Review by erikrobstad — CHANGES_REQUESTED** [review]
I have not looked at music_color_sync.c yet

File: applications/nrf5340_audio/src/modules/music_color_sync.c [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+		arm_abs_q15(output, output, (CONFIG_FFT_SAMPLE_SIZE * 2));
+		music_color_sync_fft_analyse(output, CONFIG_FFT_SAMPLE_SIZE);
+		nrfx_pwm_simple_playback(&m_pwm, &seq, 1, NRFX_PWM_FLAG_LOOP);
+		k_msleep(125);
+
+		STACK_USAGE_PRINT("music_color_sync_thread", &music_color_sync_thread_data);
+	}
+}
+
+int music_color_sync_data_update(char *pcm_data_mono, uint16_t len)
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/music_color_sync.c:116]:
```suggestion
int music_color_sync_data_update(char const * const pcm_data_mono, uint16_t len)
```

File: applications/nrf5340_audio/src/modules/music_color_sync.c [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+		k_sem_take(&music_color_sync_update_sem, K_FOREVER);
+		k_sem_take(&music_color_sync_value_sem, K_FOREVER);
+		memcpy(input, decoded_pcm_data_mono, CONFIG_FFT_SAMPLE_SIZE);
+		k_sem_give(&music_color_sync_value_sem);
+
+		arm_rfft_q15(&fft_instance, (q15_t *)decoded_pcm_data_mono, output);
+		arm_abs_q15(output, output, (CONFIG_FFT_SAMPLE_SIZE * 2));
+		music_color_sync_fft_analyse(output, CONFIG_FFT_SAMPLE_SIZE);
+		nrfx_pwm_simple_playback(&m_pwm, &seq, 1, NRFX_PWM_FLAG_LOOP);
+		k_msleep(125);
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/music_color_sync.c:110]: Why is this required? The semaphore on line 101 will likely not be available and the code will sleep there.

File: applications/nrf5340_audio/src/modules/music_color_sync.c [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+{
+	int ret;
+	static nrfx_pwm_config_t const pwm_cfg = {
+		.output_pins = {
+			MUSIC_SYNC_RED_PIN,
+			MUSIC_SYNC_GREEN_PIN,
+			MUSIC_SYNC_BLUE_PIN,
+			NRFX_PWM_PIN_NOT_USED,
+		},
+		.irq_priority	= 6,
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/music_color_sync.c:138]: NB: This is the IRQ priority and must not be mixed with the thread priority. What is the reason for 6?

File: applications/nrf5340_audio/src/modules/music_color_sync.c [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+		.base_clock		= NRF_PWM_CLK_1MHz,
+		.count_mode		= NRF_PWM_MODE_UP,
+		.top_value		= 100,
+		.load_mode		= NRF_PWM_LOAD_INDIVIDUAL,
+		.step_mode		= NRF_PWM_STEP_AUTO,
+		.skip_gpio_cfg	= false,
+		.skip_psel_cfg	= false
+	};
+
+	ret = nrfx_pwm_init(&m_pwm, &pwm_cfg, NULL, NULL);
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/music_color_sync.c:148]: This is OK, but I would think there is a Zephyr API that wraps this call to nrfx?

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 17]
```diff
@@ -104,5 +157,9 @@ config POWER_MODULE_STACK_SIZE
 	int "Stack size for power module"
 	default 1152
 
+config MUSIC_COLOR_SYNC_STACK_SIZE
+	int "Stack size for music color sync module"
+	default 2000
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/Kconfig:162]: I ran stack analyzer and got these results:
image

And added some some wiggle room for the stack, is it an appropriate estimation ?

File: applications/nrf5340_audio/src/modules/music_color_sync.c (hunk 14 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_color_sync.c:110]: This used added to slow down the refresh rate of the RGB LEDS, which gives a "smoother" change of color.

File: applications/nrf5340_audio/src/main.c [hunk 18]
```diff
@@ -233,6 +234,13 @@ void main(void)
 	ret = streamctrl_start();
 	ERR_CHK(ret);
 
+#if (CONFIG_AUDIO_DEV == HEADSET) && CONFIG_MUSIC_COLOR_SYNC
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/main.c]: I will look into this after 12.dec.

File: applications/nrf5340_audio/src/modules/music_led_sync.c [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+	while (1) {
+		k_sem_take(&music_led_sync_update_sem, K_FOREVER);
+		k_sem_take(&music_led_sync_value_sem, K_FOREVER);
+		memcpy(input, pcm_data_mono, CONFIG_FFT_SAMPLE_SIZE);
+		k_sem_give(&music_led_sync_value_sem);
+
+		arm_rfft_q15(&fft_instance, (q15_t *)pcm_data_mono, output);
+		arm_abs_q15(output, output, (CONFIG_FFT_SAMPLE_SIZE * 2));
+		music_led_sync_fft_analyse(output, CONFIG_FFT_SAMPLE_SIZE);
+		k_msleep(125);
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/music_led_sync.c:121]: Why is there an k_msleep here? Forgive me if I commented on this one before.

File: applications/nrf5340_audio/app.overlay [hunk 20]
```diff
@@ -0,0 +1,91 @@
+/ {
```
Comment (koffes) [inline applications/nrf5340_audio/app.overlay:1]: We need to re-sync on this when you are back. The boardfiles are getting a major overhaul in other PRs. See: https://github.com/zephyrproject-rtos/zephyr/pull/49720

File: applications/nrf5340_audio/src/audio/sw_codec_select.c [hunk 21]
```diff
@@ -8,9 +8,16 @@
 
 #include <zephyr/kernel.h>
 #include <errno.h>
+#include <math.h>
 
+#include "audio_sync_timer.h"
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/sw_codec_select.c]: removed

File: applications/nrf5340_audio/src/main.c [hunk 22]
```diff
@@ -232,6 +233,12 @@ void main(void)
 
 	ret = streamctrl_start();
 	ERR_CHK(ret);
+	if (CONFIG_AUDIO_DEV == HEADSET && CONFIG_MUSIC_COLOR_SYNC) {
+		ret = music_color_sync_init();
+		if (ret) {
+			LOG_ERR("Failed to init music color sync - (err: %d)", ret);
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/main.c]: Removed init form main.c

File: applications/nrf5340_audio/src/modules/Kconfig (hunk 3 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/Kconfig]: Changed from "help" to "depends on"

File: applications/nrf5340_audio/src/modules/Kconfig (hunk 5 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/Kconfig:129]: These are there so that one can change the width of the frequency bins, added config for blue LED. 
Mostly used under development of this PR, should it be removed and change the configs into "#defines"?

File: applications/nrf5340_audio/src/modules/Kconfig (hunk 6 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/Kconfig]: Changed so that they match

File: applications/nrf5340_audio/src/modules/music_color_sync.h (hunk 7 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_color_sync.h]: Added

File: applications/nrf5340_audio/src/modules/music_color_sync.h (hunk 12 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_color_sync.h]: Added small explanation

File: applications/nrf5340_audio/src/modules/Kconfig [hunk 23]
```diff
@@ -95,6 +144,10 @@ config POWER_MODULE_THREAD_PRIO
 	help
 		This is a preemptible thread
 
+config MUSIC_COLOR_SYNC_THREAD_PRIO
+	int "Priority for music color sync thread"
+	default 6
```
[the remark this answers is not in the document]
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/Kconfig]: Changed to 7

File: applications/nrf5340_audio/src/modules/music_color_sync.c (hunk 13 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_color_sync.c]: changed

File: applications/nrf5340_audio/src/modules/music_color_sync.c (hunk 16 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_color_sync.c]: Added new suggestion that uses Zephyr API

File: applications/nrf5340_audio/src/modules/music_led_sync.c (hunk 19 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/modules/music_led_sync.c:121]: This is added to give smoother transition of color for the RGB LED.

File: applications/nrf5340_audio/app.overlay [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+    aliases {
+        red-pwm-led = &pwm_led_0;
```
Comment (koffes) [inline applications/nrf5340_audio/app.overlay:3]: One new thing: We are no longer putting any "modules" into our application. That is, we want the modules to be reusable everywhere in NCS. Hence, this code should likely be put in nrf/modules or nrf/lib

File: applications/nrf5340_audio/src/modules/music_led_sync.c (hunk 19 above)
Comment (koffes) [reply applications/nrf5340_audio/src/modules/music_led_sync.c:123]: OK. Then I suggest you create a timer https://docs.zephyrproject.org/3.0.0/reference/kernel/timing/timers.html
and in the `my_expiry_function(struct k_timer *timer_id);` you can add a `k_sem_give` which releases the `while` loop

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/app.overlay (hunk 24 above)
Comment (Khuongh) [reply applications/nrf5340_audio/app.overlay:3]: Moved the code into nrf/modules. What should I do with the overlay file?

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts [hunk 25]
```diff
@@ -137,8 +137,68 @@
 
 &pwm0 {
```
Comment (Khuongh) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts:138]: Was not completely sure about the correct file for these configurations, would not mind a pointer

**Comment by Khuongh** [comment]
Moved music_led_sync from modules into lib as discussed, changed the board files to be more alike zephyrs.
@koffes @erikrobstad
